### PR TITLE
Quick Action CL Interface

### PR DESCRIPTION
**Pull request by Ace-Radom**
Create command-line interface for accessing & running Quick Actions created by the user.

## Description

This feature is based on named-pipes. A named-pipe `LenovoLegionToolkit-IPC-0` is created on LLT normal startup & closed on LLT shutdown.

The LLT process which creates this named-pipe works as a 'host', and all other processes sending Quick-Action-run-request through the pipe to the 'host' are 'clients'. In current architecture possible clients are: `Lenovo Legion Toolkit.exe` when `--run-quick-action` arg is given, and `llt.exe` when `--run` arg is given.

The purpose of `llt.exe` is avoiding UAC confirm when sending such a simple message through the pipe, which does not need admin privileges at all. I think pressing `OK` on the UAC window on each run (especially when there are multiple calls in a script) is really annoying, therefore I added this new command-line interface. The exe name is also shorter.

Although it may be a bit unnecessary, I still stored all texts in `llt.exe` in resources and made them translateable, cuz I think unlike `SpectrumTester`, `llt.exe` is an interface, not a debug/test tool, and therefore i18n support should be provided. The language of `llt.exe` is synchronised with the settings in LLT (read from `%LLT_DATAFOLDER%\lang`). (Plus if u think we don't need translated texts here, feel free to remove the resources.)

Because of the command-line interface, the installation directory would be added to the `PATH` env var of the current user. The new translations in installer were translated with DeepL (except chinese triditional & simplified, which I can understand).

## Possible Refactor

As described in #1219, I only made the IPC function for Quick Action call. If more command-line control functions are considered, IPC server & client may be moved to `LenovoLegionToolkit.Lib` and the datapack class may also be rewritten.

## TODO

- [x] IPC server & IPC client
- [x] ~~create IPC client and send quick action name when special arg is given~~ (`Lenovo Legion Toolkit.exe`) (removed)
- [x] multi-client async
- [x] close client when server has been closed
- [x] a new exe which also provides IPC client (and only), but doesn't request admin privileges (`llt.exe`)
- [x] error notification
- [x] silent run
- [x] help msg
- [ ] arg for requesting Quick Action list (on hold)
- [x] ~~set LLT install dir to sys PATH on installation~~ (will be removed due to security issues)
- [ ] ~~README update~~ (will be finished in another PR)

## Confirmed Issue / Bug

- If the named-pipe `LenovoLegionToolkit-IPC-0` has already been blocked by other processes, LLT cannot take over the control of the pipe. Although it's really rare & weird, maybe, that a named-pipe with exactly the same name has been created by other processes on the device cuz LLT ensures single instance on a normal startup, but I still added a warn message box when such thing really happens.

## Unconfirmed Issue / Bug

- As tested on two devices (R9000p 2021H & R9000p ARX8), `llt.exe` comes unresponsive on the first run after compilation. After a thought I don't think this is a bug on my side. Named-pipe had been successfully created, and it is also no longer reproducible after restarting `Lenovo Legion Toolkit.exe`. (I also cannot reproduce it after cleanup & rebuild.)

### CODE REVIEW

File: LenovoLegionToolkit.Lib.Automation/CmdLine/CmdLineIPCClient.cs
Comment (Ace-Radom) [reply LenovoLegionToolkit.Lib.Automation/CmdLine/CmdLineIPCClient.cs:97]: Nah it's okay. If the server is not found `ConnectToIPCServerAsync` will never be called.

**Comment by BartoszCichecki** [comment]
Great stuff, it's quite a bit of code, but overall looks good. I hope to have some time this weekend to review it.  🚀

**Comment by Ace-Radom** [comment]
Just realized one bug which should be really really rare (but possible). Will fix it now and I'll also start with README updates.

**Comment by Ace-Radom** [comment]
And will this change on the installer (adding install path to `PATH`) break the current upgrade flow? I'm not good at Inno Setup and don't know how this auto-upgrade works. Just being unsure if the installation folder would be added to `PATH` when user simply upgrade LLT from itself.

File: LenovoLegionToolkit.Cmdline/Program.cs
Comment (BartoszCichecki) [inline LenovoLegionToolkit.Cmdline/Program.cs:77]: Prefer early `return` over nesting:

```
if (silent)
    return;
```

File: LenovoLegionToolkit.Cmdline/Program.cs
Comment (BartoszCichecki) [inline LenovoLegionToolkit.Cmdline/Program.cs:78]: This many `if` statements would look better if replace with a `switch` statement.

File: LenovoLegionToolkit.Lib.Automation/LenovoLegionToolkit.Lib.Automation.csproj
Comment (BartoszCichecki) [inline LenovoLegionToolkit.Lib.Automation/LenovoLegionToolkit.Lib.Automation.csproj:16]: Any particular reason why protobuf is needed? Wouldn't serialization to JSON not be enough?

File: LenovoLegionToolkit.WPF/App.xaml.cs
Comment (BartoszCichecki) [inline LenovoLegionToolkit.WPF/App.xaml.cs:581]: I think the IPC server should be optional - make a setting for it.  You can place the toggle under `Integrations`.

File: LenovoLegionToolkit.WPF/App.xaml.cs
Comment (BartoszCichecki) [inline LenovoLegionToolkit.WPF/App.xaml.cs:78]: What is the purpose of this? Is it only so that the CLI can start LLT process if it's not running in background? If yes, I would remove it. It is ok to make it a requirement for LLT to run in bg for CLI to work.

File: LenovoLegionToolkit.Lib.Automation/CmdLine/IPCMessagePack.cs
Comment (BartoszCichecki) [inline LenovoLegionToolkit.Lib.Automation/CmdLine/IPCMessagePack.cs:1]: I would move or CLI related code to WPF project. It is not ideal architecture, but with the assumption that LLT needs to run in background, putting it there makes more sense if this functionality will be extended at some point.

File: LenovoLegionToolkit.Cmdline/Flags.cs
Comment (BartoszCichecki) [inline LenovoLegionToolkit.Cmdline/Flags.cs:21]: In case the CLI interface is extended, this parameter name is a bit too generic. Something like `--quickAction` would be better.

**Review by BartoszCichecki — CHANGES_REQUESTED** [review]
Good stuff overall. That is quite a bit of code, so I probably missed a thing or two ;)

1. Addid `llt.exe` was a good idea. Extracting resources is a good idea too.
2. Regarding the installer, I would skip the PATH adjustment for now, at least for the initial release. Modifying PATH is quite invasive and as you said yourself it will be tricky with the update flow.

File: LenovoLegionToolkit.Lib.Automation/CmdLine/IPCMessagePack.cs
Comment (Ace-Radom) [reply LenovoLegionToolkit.Lib.Automation/CmdLine/IPCMessagePack.cs:1]: Do u mean I should move all IPC stuff to WPF project?

File: LenovoLegionToolkit.Cmdline/Program.cs
Comment (Ace-Radom) [reply LenovoLegionToolkit.Cmdline/Program.cs:16]: I don't know a lot about C# cmd line arg parsing since I spend most of my time writing C++ codes, but as u mentioned I'll take a look, seems to be extension-friendly for the future.

File: LenovoLegionToolkit.WPF/Resources/Resource.resx
Comment (Ace-Radom) [reply LenovoLegionToolkit.WPF/Resources/Resource.resx:2165]: Yeah that's right...It's a bit too hard for non-dev users to understand. Should we add some explanation in README and let users check it by themselves?

File: LenovoLegionToolkit.Lib.Automation/CmdLine/CmdLineIPCServer.cs
Comment (Ace-Radom) [reply LenovoLegionToolkit.Lib.Automation/CmdLine/CmdLineIPCServer.cs:80]: Yeah true, if `Directory.GetFiles()` returns sth the length can only be 1.

File: LenovoLegionToolkit.Cmdline/LenovoLegionToolkit.Cmdline.csproj
Comment (Ace-Radom) [reply LenovoLegionToolkit.Cmdline/LenovoLegionToolkit.Cmdline.csproj:5]: Weird...Why tabs have been added here.

File: LenovoLegionToolkit.WPF/App.xaml.cs
Comment (Ace-Radom) [reply LenovoLegionToolkit.WPF/App.xaml.cs:78]: I didn't really understand what u meant. I put this if clause here only for judging if this is a normal LLT startup, or requesting for a Quick Action run.

File: LenovoLegionToolkit.Lib.Automation/LenovoLegionToolkit.Lib.Automation.csproj
Comment (Ace-Radom) [reply LenovoLegionToolkit.Lib.Automation/LenovoLegionToolkit.Lib.Automation.csproj:16]: I tried but there were some problems I couldn't solve, so I chose the another way. If u perfer not to link a new dependency I'll give a second try.

File: LenovoLegionToolkit.WPF/App.xaml.cs
Comment (Ace-Radom) [reply LenovoLegionToolkit.WPF/App.xaml.cs:581]: I see, will do that later.

File: LenovoLegionToolkit.WPF/Resources/Resource.resx
Comment (BartoszCichecki) [reply LenovoLegionToolkit.WPF/Resources/Resource.resx]: Yes, a readme update will be nice but let's keep it separate from this PR. You can make another one once this PR is merged with just updates for the README and I'll merge it once the release is out.

File: LenovoLegionToolkit.Lib.Automation/LenovoLegionToolkit.Lib.Automation.csproj
Comment (BartoszCichecki) [reply LenovoLegionToolkit.Lib.Automation/LenovoLegionToolkit.Lib.Automation.csproj:16]: From what I see this is only 700KB so not a big deal. Was just curious.

Comment (Ace-Radom) [reply LenovoLegionToolkit.Lib.Automation/LenovoLegionToolkit.Lib.Automation.csproj:16]: I see. But since all CLI related codes will be moved to another place I would also consider to do a minor refactor on how msgs are transfered. The current msg pack is only designed for Quick Action requests, and actually not so easy to extend. JSON is definitely a better choice but I know really few about the JSON lib LLT uses. I'll see what I can do tomorrow.

**Comment by Ace-Radom** [comment]
Things for now:

- [ ] cmd line arg parse improvements
- [ ] remove `PATH` edit on (un)install
- [ ] minor refactor on IPC datapack

**Comment by BartoszCichecki** [comment]
I had some more time to test how this works and here are couple of remarks:

1. I think there is no point in including and cmd line libs - I think we can keep CLI simple as it is now, because the idea of utilizing actions is flexible enough. Basically it will be up to the user what they want to have exposed via CLI.
2. The CLI (client) must be separated more from the rest of the code. Currently it depends on Automation and there is no point in doing that. To accomplish this you will most likely need to include a new library project (let's call it `LenovoLegionToolkit.CLI.Utils`) that will contain the data models (`IPCMessagePack`). This will need to reference protobuf too (to have access to annotations that it uses).
3. The CLI client should be moved to the CLI project, and CLI should depend only on `CLI.Utils`.
4. The CLI server, should be moved to WPF project - this is is the "host" for the named pipe, it has no place in Automation.
5. It would be good to split IPCMessagePack into two classes: one for request, one for response. Sending data objects where you have to remember which fields to set isn't very nice.
6. The argument for quick action run in WPF project is unnecessary, so let's remove that - this argument should only be available via CLI.
7. Remove PATH stuff from installer - let's let the user handle this manually for now (we could also make this an option in LLT settings to modify the path if needed, but let's wait with this - it's a minor thing and let's see what testers say).

After this is done, we can merge this into a branch here and start testing. While it's in test, I will do the rest of small adjustments.

Overall, good job 🚀

**Comment by Ace-Radom** [comment]
Thanks for the tips, first of all.

> I think there is no point in including any cmd line libs for argument parsing - I think we can keep CLI simple as it is now, because the idea of utilizing actions is flexible enough. Basically it will be up to the user what they want to have exposed via CLI.

That's true, then I'll just keep the current design and do some small improvements.

> The CLI (client) must be separated more from the rest of the code. Currently it depends on Automation and there is no point in doing that. To accomplish this you will most likely need to include a new library project (let's call it LenovoLegionToolkit.CLI.Utils) that will contain the data models (IPCMessagePack). This will need to reference protobuf too (to have access to annotations that it uses).
> The CLI client should be moved to the CLI project, and CLI should depend only on CLI.Utils.
> The CLI server, should be moved to WPF project - this is is the "host" for the named pipe, it has no place in Automation.

I see, but since the argument for Quick Action in WPF should be removed, I think there's also no need for a new lib project, cuz in this way the CLI is the only remaining client. I think I can just put it under `CLI.Utils` namespace and let it be built with CLI project. And as u said, the server will be moved to WPF project.

> Remove PATH stuff from installer - let's let the user handle this manually for now (we could also make this an option in LLT settings to modify the path if needed, but let's wait with this - it's a minor thing and let's see what testers say).

Alright, I'll remove the codes for editing `PATH` in `make_installer.iss`. But should I also remove `InnoDependencies/set_sysenv.iss` I added before? It actually works well. (I think we can just wait for the feedbacks from the testers and then see if we can remove it or not.)

**Comment by BartoszCichecki** [comment]
> But should I also remove InnoDependencies/set_sysenv.iss I added before?

I am thinking we might avoid issues with PATH if we just add it this as a switch in LLT - it will also be more "dynamic" because user will have the possibility to conveniently add/remove this without running installer.

**Comment by Ace-Radom** [comment]
FYI: I use `Lib.CLI` as the new project name since `CLI.Utils` namespace has already been used. I hope it's okay.

File: LenovoLegionToolkit.WPF/Pages/SettingsPage.xaml.cs
Comment (Ace-Radom) [reply LenovoLegionToolkit.WPF/Pages/SettingsPage.xaml.cs:647]: Nah the logic here is avoiding multi pipe servers competing for the same named-pipe.

File: LenovoLegionToolkit.Lib.CLI/CmdLineIPCClient.cs
Comment (Ace-Radom) [reply LenovoLegionToolkit.Lib.CLI/CmdLineIPCClient.cs]: Should be okay. The name of the pipe is fixed cuz a changeable name could lead to more compatibility issues so it's not considered for now.

File: LenovoLegionToolkit.Lib.CLI/Datapacks/QuickActionRequest.cs
Comment (Ace-Radom) [reply LenovoLegionToolkit.Lib.CLI/Datapacks/QuickActionRequest.cs]: `quickActionName` is not nullable so there's no need for that.

File: LenovoLegionToolkit.Lib.CLI/CmdLineIPCClient.cs
Comment (Ace-Radom) [reply LenovoLegionToolkit.Lib.CLI/CmdLineIPCClient.cs]: Good tip but in current architecture clients don't access global log function so no logs are added.

**Review by BartoszCichecki — APPROVED** [review]

**Pull request merged**
